decorators/withAuthSync/withAuthSync.js

```diff
@@ -1,40 +1,35 @@
-import React from 'react';
+import { useEffect } from 'react';
 import Router from 'next/router';
 import { authenticate } from 'common/utils/auth-utils';
 import getDisplayName from 'decorators/getDisplayName';
 
-const withAuthSync = WrappedComponent =>
-  class extends React.Component {
-    static displayName = `withAuthSync(${getDisplayName(WrappedComponent)})`;
+const syncLogout = event => {
+  if (event.key === 'logout') {
+    Router.push('/login');
+  }
+};
 
-    static async getInitialProps(ctx) {
-      const token = authenticate(ctx);
+function useWithAuthSync() {
+  useEffect(() => {
+    window.addEventListener('storage', syncLogout);
 
-      // eslint-disable-next-line unicorn/prevent-abbreviations
-      const componentProps =
-        WrappedComponent.getInitialProps && (await WrappedComponent.getInitialProps(ctx));
+    return () => {
+      window.removeEventListener('storage', syncLogout);
+      window.localStorage.removeItem('logout');
+    };
+  }, []);
+}
 
-      return { ...componentProps, token };
-    }
+useWithAuthSync.displayName = `withAuthSync(${getDisplayName(useWithAuthSync)})`;
 
-    componentDidMount() {
-      window.addEventListener('storage', this.syncLogout);
-    }
+useWithAuthSync.getInitialProps = async ctx => {
+  const token = authenticate(ctx);
 
-    componentWillUnmount() {
-      window.removeEventListener('storage', this.syncLogout);
-      window.localStorage.removeItem('logout');
-    }
-
-    syncLogout = event => {
-      if (event.key === 'logout') {
-        Router.push('/login');
-      }
-    };
+  // eslint-disable-next-line unicorn/prevent-abbreviations
+  const componentProps =
+    useWithAuthSync.getInitialProps && (await useWithAuthSync.getInitialProps(ctx));
```

Comment:
You'll wanna think about this differently if you're tryna create a hook. It looks like you're trying to do a decorator AND a hook.

`useWithAuthSync.getInitialProps` isn't the same thing as `WrappedComponent.getInitialProps`

Let's just consider `useAuthSynchronize` a method that gets called at the top of the definition of every page fn that used to be wrapped with the decorator.

Reply:
I have no idea what I was doing that day... Going to work on it today.

Reply:
You were giving it a good go!

Reply:
I've looked at one too many articles and I'm still lost, this is the best I could do
```function useWithAuthSync() {
  useEffect(() => {
    displayName(() => `withAuthSync(${getDisplayName(useWithAuthSync)})`);

    syncLogout(event => {
      if (event.key === 'logout') {
        Router.push('/login');
      }
    });

    getInitialProps(async ctx => {
      const token = authenticate(ctx);

      // eslint-disable-next-line unicorn/prevent-abbreviations
      const componentProps =
        useWithAuthSync.getInitialProps && (await useWithAuthSync.getInitialProps(ctx));

      return { ...componentProps, token };
    });

    window.addEventListener('storage', syncLogout);

    return () => {
      window.removeEventListener('storage', syncLogout);
      window.localStorage.removeItem('logout');
    };
  }, []);
}

export default useWithAuthSync;

Reply:
For one, we won't need to worry about display name anymore. This is a hook that can get called within any component - it's not longer wrapping an existing component.

You won't want to define `getInitialProps` within `useEffect` you'll want to call the `getInitialProps` property that exists upon the component calling this hook.

Does that help?

Above all else, make sure what you're writing is working! You'll want to ensure that you can't visit `/profile` unless you've logged in - for example.

Reply:
We won't be able to make this a hook https://spectrum.chat/next-js/general/static-getinitialprops-in-react-hooks~bf6f053e-3248-410e-b0bc-f200d3092b86

It needs to remain a decorator.